File: queue/Java/Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit.java
// Problem link: https://leetcode.com/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/description/?
// Time: O(n)
// Space: O(n)
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> minQ = new LinkedList<>();
        Deque<Integer> maxQ = new LinkedList<>();

        int left = 0, result = 0;

        for(int right = 0; right < nums.length; right++) {
            while(!minQ.isEmpty() && minQ.peekLast() > nums[right]) {
                minQ.pollLast();
            }

            minQ.offerLast(nums[right]);

            while(!maxQ.isEmpty() && maxQ.peekLast() < nums[right]) {
                maxQ.pollLast();
            }

            maxQ.offerLast(nums[right]);

            while(maxQ.peekFirst() - minQ.peekFirst() > limit) {
                if(maxQ.peekFirst() == nums[left])
                    maxQ.pollFirst();
                
                if(minQ.peekFirst() == nums[left])
                    minQ.pollFirst();

                left++;
            }

            result = Math.max(result, right - left + 1);
        }

        return result;
    }
}

File: queue/Python/Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit.py
# Problem link: https://leetcode.com/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/description/?
# Time: O(n)
# Space: O(n)
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        minQueue, maxQueue, left, result = deque(), deque(), 0, 0

        for right in range(len(nums)):

            while minQueue and minQueue[-1] > nums[right]:
                minQueue.pop()
            
            minQueue.append(nums[right])
            
            while maxQueue and maxQueue[-1] < nums[right]:
                maxQueue.pop()
            
            maxQueue.append(nums[right])

            while maxQueue[0] - minQueue[0] > limit:

                if nums[left] == minQueue[0]:
                    minQueue.popleft()
                
                if nums[left] == maxQueue[0]:
                    maxQueue.popleft()
                
                left += 1
            
            result = max(result, right - left + 1)
        
        return result
